### ISSUE TITLE
migrations: Slow migration crashes frontend process with a deadlock panic

### ISSUE THREAD

**Issue by mrnugget** [sourcegraph/sourcegraph#4380]
- **Sourcegraph version:** `master`/`5861f625368c7a12f00a76eb5e83995769d9cd54`

#### Steps to reproduce:

1. Create a new migration:
```
$ ./dev/add_migration.sh sleep_till_deadlock
$ ./dev/generate.sh
```
2. Open it in your editor and add the following
```sql
BEGIN;
SELECT pg_sleep(360);
COMMIT;
```
3. Start `./dev/launch.sh`

#### Expected behavior:

Sourcegraph boots up after... 360 seconds, I guess

#### Actual behavior:

`frontend` crashes with a panic (see log below in "details")

---

I'm looking into this now @slimsag, but I tagged you on this because I think you were the last one to look at this (also: deadlock detector strikes again!)

<details>

```
Starting all binaries...

14:13:56                docsite | # Doc site is available at http://localhost:5080
14:13:56                symbols | Building the /Users/thorstenball/work/sourcegraph/.bin/symbols executable...
14:13:57                  watch | Using chokidar.
14:13:57         syntect_server | Configured for production.
14:13:57         syntect_server |     => address: 0.0.0.0
14:13:57         syntect_server |     => port: 9238
14:13:57         syntect_server |     => log: critical
14:13:57         syntect_server |     => workers: 8
14:13:57         syntect_server |     => secret key: provided
14:13:57         syntect_server |     => limits: forms = 32KiB, json* = 10MiB
14:13:57         syntect_server |     => keep-alive: disabled
14:13:57         syntect_server |     => tls: disabled
14:13:57         syntect_server | Rocket has launched from http://0.0.0.0:9238
14:14:00                symbols | Building the /Users/thorstenball/work/sourcegraph/.bin/symbols executable... done
14:14:00                symbols | Building the ctags Docker image...
14:14:01                symbols | sha256:e0e1ef652374be846bd2f1e4c0a8b83cdd9e9946fdc2dec5e3b576cecec8e208
14:14:01                symbols | Building the ctags Docker image... done
14:14:13           query-runner | 2019/06/05 14:14:13 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:13           github-proxy | 2019/06/05 14:14:13 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:13           repo-updater | 2019/06/05 14:14:13 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:13               searcher | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:13               replacer | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:13              gitserver | 2019/06/05 14:14:13 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:15           query-runner | 2019/06/05 14:14:15 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:15           github-proxy | 2019/06/05 14:14:15 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:15           repo-updater | 2019/06/05 14:14:15 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:16               searcher | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:16               replacer | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:16      zoekt-indexserver | 2019/06/05 14:14:16 POST http://localhost:3090/.internal/repos/list giving up after 5 attempts
14:14:17                symbols | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:18              gitserver | 2019/06/05 14:14:18 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:18           query-runner | 2019/06/05 14:14:18 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:18           github-proxy | 2019/06/05 14:14:18 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:18           repo-updater | 2019/06/05 14:14:18 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:18               replacer | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:18               searcher | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:20                symbols | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:21              gitserver | 2019/06/05 14:14:21 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:22           query-runner | 2019/06/05 14:14:22 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:22           github-proxy | 2019/06/05 14:14:22 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:22           repo-updater | 2019/06/05 14:14:22 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:22               replacer | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:22               searcher | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:22                symbols | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23           query-runner | 2019/06/05 14:14:23 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23           repo-updater | 2019/06/05 14:14:23 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23           github-proxy | 2019/06/05 14:14:23 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23               searcher | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23               replacer | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23           query-runner | 2019/06/05 14:14:23 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23           repo-updater | 2019/06/05 14:14:23 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23           github-proxy | 2019/06/05 14:14:23 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23               replacer | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:23               searcher | received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:24              gitserver | 2019/06/05 14:14:24 received error during background config update, err: unable to fetch new configuration: Post http://localhost:3090/.internal/configuration: dial tcp [::1]:3090: connect: connection refused
14:14:26               frontend | panic: deadlock detected: you have called conf.Get or conf.Watch before the frontend has been initialized (you may need to use a goroutine)
14:14:26               frontend | goroutine 50 [running]:
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/conf.(*Store).WaitUntilInitialized(0xc0000bf720)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/conf/store.go:141 +0x33e
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/conf.(*client).Watch(0xc00028e000, 0x26e1cd8)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/conf/client.go:111 +0xf2
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/conf.Watch(0x26e1cd8)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/conf/client.go:94 +0x37
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/trace.init.0.func1()
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/trace/httptrace.go:60 +0x2d
14:14:26               frontend | created by github.com/sourcegraph/sourcegraph/pkg/trace.init.0
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/trace/httptrace.go:59 +0x132
14:14:26               frontend | goroutine 1 [IO wait]:
14:14:26               frontend | internal/poll.runtime_pollWait(0x83bdd50, 0x72, 0xffffffffffffffff)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/runtime/netpoll.go:182 +0x5e
14:14:26               frontend | internal/poll.(*pollDesc).wait(0xc000846298, 0x72, 0x2896600, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/internal/poll/fd_poll_runtime.go:87 +0x9b
14:14:26               frontend | internal/poll.(*pollDesc).waitRead(0xc000846298, 0xc00026d000, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/internal/poll/fd_poll_runtime.go:92 +0x45
14:14:26               frontend | internal/poll.(*FD).Read(0xc000846280, 0xc00026d000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/internal/poll/fd_unix.go:169 +0x39f
14:14:26               frontend | net.(*netFD).Read(0xc000846280, 0xc00026d000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/net/fd_unix.go:202 +0x7c
14:14:26               frontend | net.(*conn).Read(0xc0009724c8, 0xc00026d000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/net/net.go:177 +0xb5
14:14:26               frontend | bufio.(*Reader).Read(0xc00094d5c0, 0xc000124de0, 0x5, 0x200, 0x5, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/bufio/bufio.go:223 +0x3b1
14:14:26               frontend | io.ReadAtLeast(0x2891a40, 0xc00094d5c0, 0xc000124de0, 0x5, 0x200, 0x5, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/io/io.go:310 +0x18d
14:14:26               frontend | io.ReadFull(0x2891a40, 0xc00094d5c0, 0xc000124de0, 0x5, 0x200, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/io/io.go:329 +0xaa
14:14:26               frontend | github.com/lib/pq.(*conn).recvMessage(0xc000124dc0, 0xc0009536e0, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn.go:961 +0x1be
14:14:26               frontend | github.com/lib/pq.(*conn).recv1Buf(0xc000124dc0, 0xc0009536e0, 0xc000846200)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn.go:1011 +0x50
14:14:26               frontend | github.com/lib/pq.(*conn).recv1(0xc000124dc0, 0x0, 0xc0009536e0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn.go:1032 +0xad
14:14:26               frontend | github.com/lib/pq.(*conn).readBindResponse(0xc000124dc0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn.go:1652 +0x49
14:14:26               frontend | github.com/lib/pq.(*stmt).exec(0xc0009c4b40, 0xc0008b6b20, 0x1, 0x1)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn.go:1289 +0x3e8
14:14:26               frontend | github.com/lib/pq.(*stmt).Exec(0xc0009c4b40, 0xc0008b6b20, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn.go:1247 +0x130
14:14:26               frontend | github.com/lib/pq.(*conn).Exec(0xc000124dc0, 0x263188a, 0x1b, 0xc0008b6b20, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn.go:907 +0x409
14:14:26               frontend | github.com/lib/pq.(*conn).ExecContext(0xc000124dc0, 0x28b7e00, 0xc00004a148, 0x263188a, 0x1b, 0xc00092d890, 0x1, 0x1, 0x0, 0x0, ...)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/lib/pq@v1.0.0/conn_go18.go:43 +0x2a2
14:14:26               frontend | github.com/gchaincl/sqlhooks.(*ExecerContext).execContext(0xc00090e0d0, 0x28b7e00, 0xc00004a148, 0x263188a, 0x1b, 0xc00092d890, 0x1, 0x1, 0x0, 0x0, ...)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/gchaincl/sqlhooks@v1.1.0/sqlhooks.go:109 +0x168
14:14:26               frontend | github.com/gchaincl/sqlhooks.(*ExecerContext).ExecContext(0xc00090e0d0, 0x28b7e00, 0xc00004a148, 0x263188a, 0x1b, 0xc00092d890, 0x1, 0x1, 0x0, 0x0, ...)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/gchaincl/sqlhooks@v1.1.0/sqlhooks.go:132 +0x268
14:14:26               frontend | database/sql.ctxDriverExec(0x28b7e00, 0xc00004a148, 0x83cf748, 0xc00090e0d0, 0x0, 0x0, 0x263188a, 0x1b, 0xc00092d890, 0x1, ...)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/ctxutil.go:31 +0xe5
14:14:26               frontend | database/sql.(*DB).execDC.func2()
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:1467 +0x320
14:14:26               frontend | database/sql.withLock(0x28a2980, 0xc000846300, 0xc0009a50e8)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:3097 +0x69
14:14:26               frontend | database/sql.(*DB).execDC(0xc0009020c0, 0x28b7e00, 0xc00004a148, 0xc000846300, 0xc0008b6b00, 0x263188a, 0x1b, 0xc0009a5340, 0x1, 0x1, ...)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:1462 +0x361
14:14:26               frontend | database/sql.(*Conn).ExecContext(0xc00092d860, 0x28b7e00, 0xc00004a148, 0x263188a, 0x1b, 0xc0009a5340, 0x1, 0x1, 0x0, 0x0, ...)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:1767 +0x1ec
14:14:26               frontend | github.com/golang-migrate/migrate/v4/database/postgres.(*Postgres).Lock(0xc0009755c0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/golang-migrate/migrate/v4@v4.2.3/database/postgres/postgres.go:158 +0x346
14:14:26               frontend | github.com/golang-migrate/migrate/v4/database/postgres.(*Postgres).ensureVersionTable(0xc0009755c0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/golang-migrate/migrate/v4@v4.2.3/database/postgres/postgres.go:337 +0x4b
14:14:26               frontend | github.com/golang-migrate/migrate/v4/database/postgres.WithInstance(0xc0009020c0, 0xc00090b920, 0x0, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/github.com/golang-migrate/migrate/v4@v4.2.3/database/postgres/postgres.go:101 +0x999
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/db/dbconn.NewMigrate(0xc0009020c0, 0x0)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/db/dbconn/dbconn.go:212 +0xb3
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/db/dbconn.ConnectToDB(0x0, 0x0, 0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/db/dbconn/dbconn.go:68 +0x41c
14:14:26               frontend | github.com/sourcegraph/sourcegraph/cmd/frontend/internal/cli.Main(0x0, 0x0)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/cmd/frontend/internal/cli/serve_cmd.go:85 +0x62
14:14:26               frontend | github.com/sourcegraph/sourcegraph/cmd/frontend/shared.Main()
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/cmd/frontend/shared/frontend.go:19 +0x39
14:14:26               frontend | main.main()
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/cmd/frontend/main.go:13 +0x20
14:14:26               frontend | goroutine 39 [select]:
14:14:26               frontend | database/sql.(*DB).connectionOpener(0xc0009020c0, 0x28b7dc0, 0xc0009a6340)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:1000 +0xf1
14:14:26               frontend | created by database/sql.OpenDB
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:670 +0x29c
14:14:26               frontend | goroutine 70 [select]:
14:14:26               frontend | go.opencensus.io/stats/view.(*worker).start(0xc000211bc0)
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/go.opencensus.io@v0.19.0/stats/view/worker.go:147 +0x152
14:14:26               frontend | created by go.opencensus.io/stats/view.init.0
14:14:26               frontend |       /Users/thorstenball/code/go/pkg/mod/go.opencensus.io@v0.19.0/stats/view/worker.go:29 +0x5b
14:14:26               frontend | goroutine 105 [select]:
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/conf.(*Store).WaitUntilInitialized(0xc0000bf720)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/conf/store.go:131 +0x15f
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/conf.(*client).Watch(0xc00028e000, 0x26e11b8)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/conf/client.go:111 +0xf2
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/conf.Watch(0x26e11b8)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/conf/client.go:94 +0x37
14:14:26               frontend | github.com/sourcegraph/sourcegraph/cmd/frontend/internal/auth/userpasswd.init.1.func1()
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/cmd/frontend/internal/auth/userpasswd/config_watch.go:11 +0x2d
14:14:26               frontend | created by github.com/sourcegraph/sourcegraph/cmd/frontend/internal/auth/userpasswd.init.1
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/cmd/frontend/internal/auth/userpasswd/config_watch.go:10 +0x35
14:14:26               frontend | goroutine 107 [select]:
14:14:26               frontend | github.com/sourcegraph/sourcegraph/pkg/mutablelimiter.(*Limiter).do(0xc000394280, 0x20)
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/mutablelimiter/limiter.go:97 +0x271
14:14:26               frontend | created by github.com/sourcegraph/sourcegraph/pkg/mutablelimiter.New
14:14:26               frontend |       /Users/thorstenball/work/sourcegraph/pkg/mutablelimiter/limiter.go:41 +0x13d
14:14:26               frontend | goroutine 40 [select]:
14:14:26               frontend | database/sql.(*DB).connectionResetter(0xc0009020c0, 0x28b7dc0, 0xc0009a6340)
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:1013 +0x11f
14:14:26               frontend | created by database/sql.OpenDB
14:14:26               frontend |       /Users/thorstenball/.asdf/installs/golang/1.12.4/go/src/database/sql/sql.go:671 +0x2d2
14:14:26               frontend | Terminating frontend
14:14:26                symbols | Terminating symbols
14:14:26                  nginx | Terminating nginx
14:14:27         syntect_server | Terminating syntect_server
14:14:27      zoekt-indexserver | Terminating zoekt-indexserver
14:14:27                docsite | Terminating docsite
14:14:27              gitserver | Terminating gitserver
14:14:27        zoekt-webserver | Terminating zoekt-webserver
14:14:27 management-console-web | Terminating management-console-web
14:14:27           query-runner | Terminating query-runner
14:14:27           repo-updater | Terminating repo-updater
14:14:27               searcher | Terminating searcher
14:14:27     management-console | Terminating management-console
14:14:27               replacer | Terminating replacer
14:14:27           github-proxy | Terminating github-proxy
14:14:27                  watch | Terminating watch
goreman: exit status 2
```
</details>

**Comment by slimsag** [comment]
Ah apologies I saw this last night and meant to respond.

The deadlock detector here is time-based. If the frontend doesn't start in 30s, it assumes there is a deadlock and panics. Obviously, it being time-based isn't optimal because in cases like this it hides the fact that the real cause here is something else (it is picking up the fact that your frontend isn't starting after 30s and saying "well there must be a deadlock then!"). However, we can't do anything to make it not time-based.

One thing we can and should do is improve the wording here to make it more clear that this isn't guaranteed. For example, that panic should instead read:

> panic: potential deadlock detected: the frontend's configuration server hasn't started after 30s indicating a deadlock may be happening. A common cause of this is calling conf.Get or conf.Watch before the frontend has started fully (e.g. inside an init function) and if that is the case you may need to invoke those functions in a separate goroutine.

More verbose, but also much less ambiguous in what it is actually definitively telling you.

Now, as far as the actual timeout duration goes, we use 30s in dev because:

1. If your dev server doesn't start in 30s something is wrong
2. If we made the timeout longer than 30s, when a deadlock did occur nobody would know about the race detector that helps them and instead would kill the process and struggle to debug it some other way (is the thought process).

For your purposes here of reproducing a lengthy DB migration, it would make sense for us to add an env var that can disable it. i.e. `CONF_DEADLOCK_DETECTOR=false ./enterprise/dev/start.sh`

In non-dev environments the timeout is much less aggressive, 5 minutes. I chose 5 minutes purely as a guess with that being the longest amount of time a migration / DB initialization should take to occur. If it is longer than that, we should make it even more relaxed. We control these here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@b05101977f8a29534c743bc3aeda5c87acecf5dd/-/blob/pkg/conf/store.go#L126-129
